rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_equipment_associated_with_each_swh_distribution_system.py

```diff
@@ -0,0 +1,86 @@
+
+    Parameters
+    ----------
+    rmd: dict, RMD at RuleSetModelDescription level
+
+    Returns
+    -------
+    swh_and_equip_dict: A dictionary containing where the keys are the SWH Distribution System IDs and values are dictionaries where keys are the type of SWH equipment and values are the ids of the connected equipment.
+                        Example:  {"swh_distribution1":{"swh_heating_eq":["swh_eq1","swh_eq2"], "pumps":["p1"], "tanks":["t1"], "piping":["piping1"], "solar_thermal":[], "uses":["sp1_use","sp2_use"], "spaces_served":[sp1,sp2]}}
+
+    """
+
+    swh_and_equip_dict = {}
+    for distribution in find_all(
+        "$.service_water_heating_distribution_systems[*]", rmd
+    ):
+        swh_and_equip_dict[distribution["id"]] = SWHDistributionAssociations()
+        swh_and_equip_dict[distribution["id"]].tanks = [
+            tank["id"] for tank in find_all("$.tanks[*]", distribution)
+        ]
+        for piping in distribution.get("service_water_piping", []):
+            queue = deque([piping])
+            piping_ids = []
+            # BFS approach
+            while queue:
+                current_piping = queue.popleft()
+                piping_ids.append(current_piping["id"])
+                children = current_piping.get("child", [])
+                queue.extend(children)
+            swh_and_equip_dict[distribution["id"]].piping.extend(piping_ids)
+
+        for pump in find_all("$.pumps[*]", rmd):
+            if (
+                pump.get("loop_or_piping")
+                in swh_and_equip_dict[distribution["id"]].piping
+            ):
+                swh_and_equip_dict[distribution["id"]].pumps.append(pump["id"])
+
+        for swh_equip in find_all("$.service_water_heating_equipment[*]", rmd):
+            if swh_equip.get("distribution_system") == distribution["id"]:
+                swh_and_equip_dict[distribution["id"]].swh_heating_eq.append(
+                    swh_equip["id"]
+                )
+                for solar_t in swh_equip.get("solar_thermal_systems", []):
+                    swh_and_equip_dict[distribution["id"]].solar_thermal.append(
+                        solar_t["id"]
+                    )
+
+    # TODO revise the json path if the service_water_heating_uses is relocated in the schema
+    for swh_use in find_all(
+        "$.buildings[*].building_segments[*].zones[*].spaces[*].service_water_heating_uses[*]",
+        rmd,
+    ):
+        distribution_id = swh_use.get("served_by_distribution_system")
+        if distribution_id:
+            swh_and_equip_dict[distribution_id].uses.append(swh_use["id"])
+            for space_id in get_spaces_served_by_swh_use(rmd, swh_use["id"]):
+                if space_id not in swh_and_equip_dict[distribution_id].spaces_served:
```

Comment:
It is OK to implement in this way but here is my take: Looking at the complexity, this code block start with `for swh_use in find_all(...)`, then loop into `for space_id in space_ids_served_by_swh_use`, and lastly, the `if space_id not in swh_and_equip_dict`. The complexity is `O(n^3)`. Note, the `in` operation in Python has a complexity of `O` when running on a list.

A better to do it is:
```python
if distribution_id:
            swh_and_equip_dict[distribution_id].uses.append(swh_use["id"])
            for space_id in get_spaces_served_by_swh_use(rmd, swh_use["id"]):
                swh_and_equip_dict[distribution_id].spaces_served.append(space_id)
            swh_and_equip_dict[distribution_id].spaces_served = list(set(swh_and_equip_dict[distribution_id]))
```
In this case, although the `list(set())` complexity is `O(n)`, but because it is not nested inside the `for space_id` loop, so it does not contribute to the overall complexity - the code block complexity is still O(n^2).

It is not a major issue but thought can be helpful to share.